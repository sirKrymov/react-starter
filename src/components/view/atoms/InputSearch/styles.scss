@import '../../../../styles/configs';
@import '../../../../styles/mixins';

.input-search {
  min-height: inherit;
  position: relative;

  input {
    -webkit-appearance: none;
    background-color: var(--bgPrimary);
    border-radius: $border-radius;
    transition: border-color $anim-time-fast, background-color $anim-time-fast,
      opacity $anim-time;
    box-shadow: none;
    min-height: inherit;
    padding: 0 16px;
    outline: none;
    border: 1px solid var(--borderPrimary);
    color: var(--textPrimary);
    width: 100%;

    &::placeholder {
      color: var(--textHint);
    }

    &:focus {
      border-color: var(--thirdy);
      box-shadow: none;
    }
  }

  &__icon-wrapper {
    transform: translateY(-50%);
    position: absolute;
    height: 100%;
    left: 0;
    top: 50%;
    @include center;

    &--xs {
      width: 28px;
    }

    &--sm {
      width: 36px;
    }

    &--md {
      width: 44px;
    }

    &--lg {
      width: 52px;
    }
  }

  &__clear-icon-wrapper {
    transform: translateY(-50%);
    position: absolute;
    height: 100%;
    right: 0;
    top: 50%;
    @include center;

    &--xs {
      width: 28px;
    }

    &--sm {
      width: 36px;
    }

    &--md {
      width: 44px;
    }

    &--lg {
      width: 52px;
    }
  }

  &--icon-wrapper-xs {
    input {
      padding-left: 28px;
    }
  }

  &--icon-wrapper-sm {
    input {
      padding-left: 36px;
    }
  }

  &--icon-wrapper-md {
    input {
      padding-left: 44px;
    }
  }

  &--icon-wrapper-lg {
    input {
      padding-left: 52px;
    }
  }

  &--clear-wrapper-xs {
    input {
      padding-right: 28px;
    }
  }

  &--clear-wrapper-sm {
    input {
      padding-right: 36px;
    }
  }

  &--clear-wrapper-md {
    input {
      padding-right: 44px;
    }
  }

  &--clear-wrapper-lg {
    input {
      padding-right: 52px;
    }
  }
}
