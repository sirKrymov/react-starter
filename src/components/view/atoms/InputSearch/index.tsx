// @SETUP: Please don't forget to remove this component or current comment if component is used in project
import React, { ChangeEvent, ReactElement } from 'react';

import { IoMdCloseCircle, IoIosSearch } from 'react-icons/io';
import cn from 'classnames';

import { ReactIcon } from 'components/view/atoms/ReactIcon';
import { UiSize } from 'components/view/atoms/UiSize';

import './styles.scss';

interface IProps {
  onChange(e: ChangeEvent<any>): void;
  onIconClick?(): void;
  placeholder?: string;
  className?: string;
  onClear?(): void;
  height?: 'xs' | 'sm' | 'md' | 'lg' | 'full';
  width?: 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'full' | 'content';
  value: string;
  name: string;
  id: string;
}

export function InputSearch({
  placeholder = 'Search...',
  onIconClick,
  className,
  onChange,
  onClear,
  height = 'md',
  width,
  value,
  name,
  id
}: IProps): ReactElement {
  const inputSearchClass = cn(
    {
      'input-search': true,
      [`input-search--icon-wrapper-${height}`]: height,
      [`input-search--clear-wrapper-${height}`]: height && onClear
    },
    className
  );

  const iconWrapperClass = cn({
    'input-search__icon-wrapper': true,
    [`input-search__icon-wrapper--${height}`]: height
  });

  const clearWrapperClass = cn({
    'input-search__clear-icon-wrapper': true,
    [`input-search__clear-icon-wrapper--${height}`]: height
  });

  return (
    <UiSize width={width} height={height}>
      <div className={inputSearchClass}>
        <div
          className={iconWrapperClass}
          onClick={onIconClick}
          style={{ ...(onIconClick && { cursor: 'pointer' }) }}
        >
          <ReactIcon>
            <IoIosSearch />
          </ReactIcon>
        </div>

        <input
          autoComplete="off"
          placeholder={placeholder}
          onChange={onChange}
          value={value}
          type="text"
          name={name}
          id={id}
        />

        {value && onClear && (
          <div className={clearWrapperClass} onClick={onClear}>
            <ReactIcon size="md">
              <IoMdCloseCircle />
            </ReactIcon>
          </div>
        )}
      </div>
    </UiSize>
  );
}
