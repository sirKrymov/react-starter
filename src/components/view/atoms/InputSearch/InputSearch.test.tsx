import React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';

import { InputSearch } from '../InputSearch';

describe('<InputSearch> atom component', () => {
  const handleClick = sinon.spy();
  const handleChange = jest.fn();

  let shallowComponent: ShallowWrapper;

  const createComponent = ({
    onChange,
    value,
    ...rest
  }: Record<string, any>) => {
    shallowComponent = shallow(
      <InputSearch
        onChange={onChange}
        value={value}
        name="search"
        id="check-search"
        {...rest}
      />
    );
  };

  describe('WHEN the component loads', () => {
    beforeEach(() => {
      createComponent({ onChange: handleChange, value: '' });
    });

    it('THEN it should render without crashing', () => {
      expect(shallowComponent.html()).not.toHaveLength(0);
    });

    it('THEN it should match snapshot', () => {
      expect(shallowToJson(shallowComponent)).toMatchSnapshot();
    });

    it('THEN it should fires on change correctly', () => {
      const event = { target: { value: 'search text' } };

      expect(shallowComponent.find('input').simulate('change', event));
      expect(handleChange).toHaveBeenCalledTimes(1);
      expect(handleChange).toHaveBeenLastCalledWith(event);
    });
  });

  describe('WHEN the component have different props', () => {
    it('WHEN to component have length and value="search"', () => {
      createComponent({
        onChange: handleChange,
        value: 'search',
        onClear: handleClick
      });

      expect(
        shallowComponent.find('div.input-search__clear-icon-wrapper')
      ).toHaveLength(1);
    });

    it('WHEN to component have onIconClick', () => {
      createComponent({
        onChange: handleChange,
        value: 'search',
        onIconClick: true
      });

      expect(
        shallowComponent.find('div.input-search__icon-wrapper').prop('style')
      ).toEqual({ cursor: 'pointer' });
    });
  });
});
