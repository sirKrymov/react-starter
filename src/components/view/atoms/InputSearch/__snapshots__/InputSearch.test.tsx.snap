// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputSearch> atom component WHEN the component loads THEN it should match snapshot 1`] = `
<UiSize
  height="md"
>
  <div
    className="input-search input-search--icon-wrapper-md"
  >
    <div
      className="input-search__icon-wrapper input-search__icon-wrapper--md"
      style={Object {}}
    >
      <ReactIcon>
        <IoIosSearch />
      </ReactIcon>
    </div>
    <input
      autoComplete="off"
      id="check-search"
      name="search"
      onChange={[MockFunction]}
      placeholder="Search..."
      type="text"
      value=""
    />
  </div>
</UiSize>
`;
